presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Optional;
+
+public interface Rule
+{
+    Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator);
```

Comment:
I wonder if there is a case when `Rule` might need access to `Memo`. 

For example a rule which tries to identify common subplans and replace each of their occurrence with the same `GroupNode`, then such rule needs an access to the whole plan. Do you think that such rule should recursively call `lookup`  in order to traverse the plan in the current design?